kernel/EigenSolvers/JacobiDavidsonMaxwell.cpp

```diff
@@ -0,0 +1,889 @@
+        this->eta = (this->iter > 0 && eps < 0.5) ? rho : this->tau;
+
+        // copy res to res_new and left project
+        ierr = VecCopy(this->residue_vect, residue_vect_copy);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        computeProjection(residue_vect_copy, this->Qt);
+
+        // solve correction equation
+        solveCorrectionEquation(residue_vect_copy, correction_vect);
+
+        // ortho correction vector
+        computeProjection(correction_vect, this->Qt);
+
+        // project correction vector
+        projectCorrectionVector(correction_vect);
+
+        // create search vector
+        ierr = VecCopy(correction_vect, search_vect);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        computeProjection(search_vect, this->V);
+        normalizeVector(search_vect);
+
+        k++;
+
+        // Add column to V
+        ierr = BVInsertVec(this->V, this->V_current_size, search_vect);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+        this->V_current_size++;
+        ierr = BVSetActiveColumns(this->V, 0, this->V_current_size);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        // compute eigenpairs on small subspace
+        ierr = computeSmallEigenvalues(small_evals, small_evects);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        // create q
+        ierr = VecDuplicate(this->residue_vect, &q);
+        CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+        found = PETSC_TRUE;
+        idx_evect = 0;
+        while (found) {
+
+            // rho = Lk[0]
+            rho = small_evals[0];
+
+            // q = V sk0
+            ierr = VecGetArray(small_evects[0], &vec_values);
+            CHKERRABORT(PETSC_COMM_WORLD, ierr);
+            ierr = BVMultVec(this->V, 1.0, 0.0, q, vec_values);
+            CHKERRABORT(PETSC_COMM_WORLD, ierr);
+            ierr = VecRestoreArray(small_evects[0], &vec_values);
+            CHKERRABORT(PETSC_COMM_WORLD, ierr);
+
+            // r = A q - rho M q
+            computeResidueVector(q, rho, this->residue_vect);
+
+            // compute residue threshold
+            computeThreshold(q, this->residue_vect, &eps);
+
+            // compute covnergence criteria
+            found = (eps < this->tolerance && k > 0) ? PETSC_TRUE : PETSC_FALSE;
+            if (found) {
+                logger(VERBOSE,
+                       "JacobiDavidsonSolver new eigenvector found with "
+                       "eigenvalue %",
+                       rho);
+                // store the eigenvalue/eigenvector
+                this->eigenvalues.push_back(rho);
+
+                // add a new vector to the basis
+                ierr = BVInsertVec(Q, Q_current_size, q);
+                CHKERRABORT(PETSC_COMM_WORLD, ierr);
+                this->Q_current_size++;
+
+                if (this->Q_current_size >= nev) {
+                    stop = PETSC_TRUE;
+                    break;
+                }
+
+                // Compute [tmp_v1 .... tmp_vk-1] = V [s2, .... sk]
```

Comment:
I think that quite a bit of this juggling with vectors might have been simplified by sharing a BV for [Q, V]. By using BVSetActiveColumns one can then select either the Q or V part. What this would do to the readability is a second question though.